clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -335,6 +335,14 @@ public void poll(long timeout) throws IOException {
         // we use the time at the end of select to ensure that we don't close any connections that
         // have just been processed in pollSelectionKeys
         maybeCloseOldestConnection(endSelect);
+
+        Iterator<String> connectedIt = connected.iterator();
+        while (connectedIt.hasNext()) {
+            if (disconnected.containsKey(connectedIt.next())) {
+                log.warn("Channel {} is marked connected and disconnected at the same time");
+                connectedIt.remove();
```

Comment:
Maybe we could keep it simple and add a test for `Selector.pollSelectedKeys` where `SelectionKey` and the channel returned by the `SelectionKey` are mocked. What do you think?